Allow `rack >= 3` in Rails.

ioquatix opened the pull request:
Allow Rack 3 to be used in Rails.

## Blocked By:

- https://github.com/rails/rails/pull/46206

casperisfine commented:
You'll have to bundle update `rack` to see if the test suite passes.

As for the failure you see, it was fixed on `main`, so you can rebase.

p8 commented:
Wouldn't this also allow rack 4?

simi commented on Gemfile.lock:
```diff
@@ -321,10 +322,6 @@ GEM
     nokogiri (1.13.3)
       mini_portile2 (~> 2.8.0)
       racc (~> 1.4)
-    nokogiri (1.13.3-x86_64-darwin)
```
Any idea why this was removed?

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -422,8 +422,22 @@ def clear(options = {})
       def write(headers)
-        if header = make_set_cookie_header(headers[HTTP_HEADER])
-          headers[HTTP_HEADER] = header
+        # Make a fake response for the cookie to write to:
```
```suggestion
        # Make a fake response for the cookie to write to.
```

ioquatix replied:
I always use colon when referring to the next line rather than just making a statement about execution. Is that a problem?

rafaelfranca commented on actionpack/actionpack.gemspec:
```diff
@@ -35,7 +35,8 @@ Gem::Specification.new do |s|
 
-  s.add_dependency "rack",      "~> 2.0", ">= 2.2.4"
+  s.add_dependency "rack", ">= 2.2.4", "< 4"
+  s.add_dependency "rack-session"
```
How does this work with rack 2? How are we use that the code in rack-session is the one that is being required when we `require "rack/session/abstract/id"? Doesn't it depends on the order those gems are in the load path?

ioquatix replied:
The current release of `rack-session` IIRC is essentially line-for-line file-for-file compatible so it's okay to load it with Rack 2. However, I imagine this will eventually change, so the plan is this:

1. Release a stable 1.x release that is compatible with the current interface and should be okay to have loaded with Rack 2.
2. Release a stable 2.x release that is only for Rack >= 3.0 when that interface is changed in any substantial way.

ioquatix commented:
On my local dev machine, just considering `actionpack`, I've got the test suite down to 27 failures and 2 errors.

ioquatix commented:
One thing I could use some help with is adding Rack 2 and Rack 3 to the test matrix. It is probably only strictly required for `actionpack` but maybe others wouldn't hurt.

ioquatix commented:
Okay, zero failures locally now.

ioquatix commented:
I've released `rackup` and `rack-session` with v1 (compatible with rack 2) and v2 (compatible with rack 3+). That should make it easier declaring dependencies in Rails.

ioquatix commented:
I can't get the 2.7 railties test suite to fail locally, I have no idea what's wrong, maybe there is a (BuildKite specific) cached bundle is causing issues?

rafaelfranca commented on actionpack/actionpack.gemspec:
```diff
@@ -35,7 +35,9 @@ Gem::Specification.new do |s|
   s.add_dependency "activesupport", version
 
-  s.add_dependency "rack",      "~> 2.0", ">= 2.2.4"
+  s.add_dependency "rack", ">= 2.2.4", "< 3"
```
```suggestion
  s.add_dependency "rack", ">= 2.2.4"
```

rafaelfranca commented on actionpack/actionpack.gemspec:
```diff
@@ -35,7 +35,9 @@ Gem::Specification.new do |s|
 
-  s.add_dependency "rack",      "~> 2.0", ">= 2.2.4"
+  s.add_dependency "rack", ">= 2.2.4", "< 3"
+  # s.add_dependency "rack", ">= 2.2.4", "< 4"
```
```suggestion
```

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -208,6 +208,55 @@ class Cookies
+    # Example:
+    #   set_cookies = headers["set-cookie"]
+    #   parse_set_cookies_headers(set_cookies)
+    def self.parse_set_cookies_headers(set_cookies)
```
Why we need to parse cookies now?

ioquatix replied:
A ton of tests depend on it and they were just using string comparisons which is insufficient. I don't know where else to put this, can you suggest a better place? Somewhere in the test helpers?

ioquatix replied:
I've moved this code into the test helpers, hopefully that's a better place for it.

ioquatix commented:
Well, that was unexpected, it suddenly started passing!

ioquatix commented:
I've made a companion PR which shows these changes working on Rack 2: https://github.com/rails/rails/pull/47052

ioquatix commented:
Okay, I'm fairly happy with this PR, can I please get some more reviews/feedback?

- Passing on Rack 2 and Rack 3.
- Minimal changes - mostly ones which align and defer functionality to Rack.
- Most test changes are for compatibility and test the identical behaviour as before.

We can start thinking about bigger changes for Rails 8 which I believe will be a good opportunity to go all in with Rack 3 and drop Rack 2.

matthewd commented:
Unless I'm misreading, this is proposing to make (Rack 2 using) applications have two gems on their load path providing an overlapping set of files. 

In that situation, which will win? 

In the best possible case I can imagine, it's globally deterministic (e.g. alpha ordering of gem names), and the only issue is that users might believe they've updated to avoid a security issue while actually running code from the other one.

In the more probable case, it might be determined by order of entries in the Gemfile, making remote debugging of an application troublesome.

In the worst possible (but I hope less likely) case, it might be determined by local filesystem ordering, meaning two copies of the same application (or e.g. at least two copies running on different operating systems) might be loading different files out of a locked set of dependencies.

(Does this change when using one of bundler's vendoring techniques? Forgoing bundler entirely? Using rubygems's own built-in Gemfile support?)

None of those seem like compellingly safe upgrade paths to me, even if we do stipulate that the two copies are intended to be essentially identical for suitably chosen pairs.

ioquatix commented:
@matthewd the v1 releases are empty (for Rack v2) and the v2 which have files are Rack v3+ only.

jeremyevans commented on actionpack/lib/action_dispatch/http/response.rb:
```diff
@@ -5,6 +5,14 @@
+  require "rack/headers"
+rescue LoadError
+  # For `Rack::Utils::HeaderHash`:
+  require "rack/utils"
```
If acceptable by the style guide, I would move this to the `Headers` definition, so you don't need two conditionals.

ioquatix replied:
Implemented.

jeremyevans commented on actionpack/lib/action_dispatch/http/response.rb:
```diff
@@ -73,11 +64,11 @@ def to_hash
+    alias_method :header, :headers
 
-    delegate :[], :[]=, to: :@header
+    delegate :[], :[]=, to: :@headers
```
These `header`/`headers` local and instance variable and and `Header`/`Headers` constant name changes don't seem to be necessary for the PR.  Not opposed to them eventually, but this PR is large enough without them. It's best to not mix unnecessary stylistic changes with structural changes to add new features, IMO, as it makes review significantly more challenging.

ioquatix replied:
I agree, so I removed it and will re-roll this later.

jeremyevans commented on actionpack/lib/action_dispatch/http/response.rb:
```diff
@@ -168,10 +163,16 @@ def self.merge_default_headers(original, default)
+      @headers = Headers.new
+
+      if headers
+        headers.each do |key, value|
```
`headers&.each` assuming that is allowed by the style guide.

ioquatix replied:
Thanks, I've updated the relevant PR, that is good feedback.

jeremyevans reviewed (commented):
@ioquatix asked me to review the `ActionDispatch::Response` changes, so I've provided my feedback limited to those changes.

ioquatix commented:
To simplify this PR, I'm splitting out some parts:

- https://github.com/rails/rails/pull/47066 (cookie assertions only).

ioquatix committed:
Don't assume post with params will preserve input body.

Rack 3 drops the requirement for input bodies to be rewindable, and
params, once read, may not be available for "raw_post". Update the tests
to prefer to use raw "body:" rather than "params:" which can be consumed
by Rack::Request in Rack 3.

ioquatix commented:
Okay, I've split out most of the pre-cursor PRs, most of them are trivial, hopefully that makes review easier!

cc @tenderlove

matthewd commented:
Ah okay, thanks for clarifying that @ioquatix -- I'd previously seen the 0.x releases, and was obviously too casual in my check that the 1.x series still had a `lib/rack` without noting it's actually just the version file.

ioquatix commented:
Thank you everyone for your quick turn around time reviewing these PRs, I really appreciate it!!!

ioquatix commented:
In theory this introduces a path to fix the following issues:

- https://github.com/rails/rails/issues/14358

ioquatix commented:
I've rebased this PR on all the merged changes thus far.

ioquatix commented:
I am rebasing this on top of the latest CI changes.

ioquatix commented:
Thanks to @zzak's work, we now have integrated Rack v2 and Rack v3 tests in the same matrix.

ioquatix commented:
Team, there are only a few PRs left, most of them just require someone to make a decision. I'm happy to make any required changes.

- https://github.com/rails/rails/pull/47076 - test changes only.
- https://github.com/rails/rails/pull/47078 - not strictly required but prevents unnecessary warnings.
- https://github.com/rails/rails/pull/47080 - no test changes, retains compatibility with Rack 2.
- https://github.com/rails/rails/pull/47092 - required for Rack 3.
- https://github.com/rails/rails/pull/47147 - not strictly required but good for consistency.

jrochkind commented:
Hi, I hope this is an appropriate place for this question, please let me know if not. 

> we now have integrated Rack v2 and Rack v3 tests in the same matrix.

I run some of my CI against rails edge just for early warning. Is rails edge currently meant to support rack 2 _or_ 3, as in:  it should be possible to run rails edge with rack 2?

I am having trouble getting bundler to do a resolution for rails edge and rack 2. 

Current edge actionpack gemspec requires `"rack-session", ">= 1.0.1"`.

https://github.com/rails/rails/blob/fee14347348ad99ff6260225aaebc8809a983fa9/actionpack/actionpack.gemspec#L39

There does not seem to be a `rack-session` 1.0.1, the lowest `rack-session` version that is `>= 1.0.1` seems to be 2.0, which requires `rack >= 3`. https://rubygems.org/gems/rackup/versions/2.0.0

So this seems to make current edge rails -- by indirect dependencies -- require rack 3, the dependency tree can't be resolved with rack 2. 

Or, wait, ok... is there an unreleased `rack-session` 1.0.1  that you are somehow testing Rails against, to test against rack 2, and required to run rails-edge with rack 2?  Oddly, while there is a [1.0.1 tag in the rack session repo](https://github.com/rack/rack-session/tree/v1.0.1)... there is not a 1.0.1 released to rubygems (or mentioned in [rack-session github releases](https://github.com/rack/rack-session/releases))?   

I thought I was helpfully pointing out that current edge rails could not be resolved against rack 2, which was not what you intended... but now I'm thinking maybe there's a whole plan here involving yet-to-be-released rack-session 1.0.1 that I'm just not privy to? Anyway, any advice on how to run my own CI against rails-edge but with rack 2 would be welcome. (because other of my dependencies may not yet be compatible with rack 3).

ioquatix commented:
Some changes were made in https://github.com/rails/rails/pull/47080 so I've rebased this PR on that one.

drnic commented:
Great talk at RubyConfAU today @ioquatix! Looking forward to my rails 7.1 app being more awesome. Thanks for all the work!

ioquatix commented:
A related issue: https://github.com/rails/rails/issues/47456

ioquatix commented:
## Outstanding PRs:

- https://github.com/rails/rails/pull/47076 - test changes only, but still tests the same functionality/method.
- https://github.com/rails/rails/pull/47078 - not strictly required but prevents unnecessary warnings.
- https://github.com/rails/rails/pull/47477

ioquatix commented:
## Outstanding PRs

- https://github.com/rails/rails/pull/47076
- https://github.com/rails/rails/pull/47078
- https://github.com/rails/rails/pull/47652

JoeDupuis commented on actionpack/test/controller/test_case_test.rb:
```diff
@@ -262,11 +262,11 @@ def to_param
+    body = "Hello World"
 
-    post :render_body, params: params.dup
+    post :render_body, body: body
```
I don't think setting the body directly is the right fix here. 

This would [bypass the `body_stream` branch](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/http/request.rb#L356-L362) in `ActionDispatch::Request#body` and [there is already have a test for it.](https://github.com/rails/rails/blob/main/actionpack/test/controller/test_case_test.rb#L277-L280)

This is currently failing on CI, so I tried a different fix and submitted a [PR](https://github.com/rails/rails/pull/48158) targeting main to bring back CI to green.

JoeDupuis commented on actionpack/test/controller/test_case_test.rb:
```diff
@@ -235,10 +235,10 @@ def test_url_options_reset
+    post :render_raw_post, body: body
 
-    assert_equal params.to_query, @response.body
+    assert_equal body, @response.body
```
This test pass for me on main with Rack 3.0 
Unless I am missing something, I don't think we need to change it anymore.

ioquatix commented:
Okay, as far as I'm concerned, this is now all working the best it can be for a 7.1 release. Thanks everyone for your huge effort and patience. It was truely a multi-year project.

I'd also recommend we make the rack-3 parts of the CI non-optional if they aren't already.

ioquatix commented:
Summary of all work done:


- https://github.com/rails/rails/pull/47066
- https://github.com/rails/rails/pull/47073
- https://github.com/rails/rails/pull/47074
- https://github.com/rails/rails/pull/47075
- https://github.com/rails/rails/pull/47076
- https://github.com/rails/rails/pull/47077
- https://github.com/rails/rails/pull/47078
- https://github.com/rails/rails/pull/47079
- https://github.com/rails/rails/pull/47080
- https://github.com/rails/rails/pull/47081
- https://github.com/rails/rails/pull/47082
- https://github.com/rails/rails/pull/47083
- https://github.com/rails/rails/pull/47085
- https://github.com/rails/rails/pull/47086
- https://github.com/rails/rails/pull/47091
- https://github.com/rails/rails/pull/47092
- https://github.com/rails/rails/pull/47094
- https://github.com/rails/rails/pull/47147
- https://github.com/rails/rails/pull/47257
- https://github.com/rails/rails/pull/47347
- https://github.com/rails/rails/issues/47561
- https://github.com/rails/rails/pull/47652
- https://github.com/rails/rails/pull/47885
- https://github.com/rails/rails/pull/48158
- https://github.com/rails/rails/pull/48437
- https://github.com/rails/rails/pull/48455

piradata commented:
The last version of actionpack (7.0.6) still wants rack to be ~> 2.0, >= 2.2.4.
Do you guys know if there is any open PR on actionpack about changing this?

skipkayhil commented:
> The last version of actionpack (7.0.6) still wants rack to be ~> 2.0, >= 2.2.4. Do you guys know if there is any open PR on actionpack about changing this?

The work to support Rack 3 has been done on the `main` branch, so it will be released as Rails 7.1
